runtime/lua/vim/Iter.lua

```diff
@@ -0,0 +1,332 @@
+--- Iterator implementation.
+
+---@class Iter
+---@field fn function
+---@field head ?number
+---@field tail ?number
+local Iter = {}
+
+Iter.__index = Iter
+
+Iter.__call = function(self)
+  return self:next()
+end
+
+--- Add a filter/map step to the iterator.
+---
+--- Example:
+--- <pre>
+--- > local it = vim.iter({ 1, 2, 3, 4 }):filter_map(function(i, v)
+--- >   if v % 2 == 0 then
+--- >     return i, v * 3
+--- >   end
+--- > end)
+--- > it:collect()
+--- { 6, 12 }
+--- </pre>
+---
+---@param f function(...):any Mapping function. Takes all values returned from the previous stage
+---                            in the pipeline as arguments and returns a new value. Nil values
+---                            returned from `f` are filtered from the output.
+---@return Iter
+function Iter.filter_map(self, f)
```

Comment:
What's the reason to merge map and filter into one? In pretty much any language/stdlib these are separate functions.

Reply:
Some languages have [transducers](https://juliafolds.github.io/Transducers.jl/dev/explanation/comparison_to_iterators/#comparison-to-iterators), too :P

Reply:
On further thought I think `filter()` is worth having, so I brought it back.

Reply:
> In Lua, we cannot have nil values in the middle of a table,

That's not strictly speaking correct; `{ 1, 2, nil, 4 }` is a perfectly valid table (and array but not list).